compiler/ast/checked_ast.nim

```diff
@@ -0,0 +1,112 @@
+## The ``CheckedAst`` type is implemented here, along with its accessor
+## routines and routines for creating and manipulating ``CheckedAst``
+## instances.
+##
+## ``CheckedAst`` is a wrapper around ``PNode``, and is intended to make
+## propagation less manual, by keeping track of whether there is an error
+## *somewhere* in the managed AST.
+##
+## The idea is to use ``CheckedAst`` as the return type of all semantic
+## analysis procedures, eventually removing the need for wrapper errors
+## (``adWrappedError``).
+
+import
+  compiler/ast/[
+    ast,
+    ast_types,
+    errorhandling,
+    lineinfos
+  ],
+  compiler/front/[
+    options
+  ]
+
+type
+  CheckedAst* = object
+    ## Represents either in-progress or fully semanticized AST.
+    n: PNode       ## stores the AST. Never nil
+    hasError: bool ## tracks whether `n` contains an ``nkError`` node
+                   ## somewhere
+
+proc prepare*(x: var CheckedAst, n: PNode) =
+  ## Initialize `x` as a node of kind ``n.kind``. Space for all children of
+  ## `n` is allocated, but the child nodes are not copied. For the flags, only
+  ## the persistent ones are copied.
+  assert x.n == nil, "already initialized"
+  assert n.kind != nkError
+  x.n = shallowCopy(n)
+
+func initWith*(x: var CheckedAst, n: sink PNode) =
+  ## Initializes `x`, using `n` as the tree. This is only meant to be used
+  ## with the direct result of a ``newNodeX`` or ``newTreeX`` call.
+  assert x.n == nil, "already initialized"
+  assert n.kind != nkError
+  x.n = n
+
+proc newError*(c: ConfigRef, n: sink CheckedAst, diag: PAstDiag): CheckedAst =
+  ## Creates an ``CheckedAst`` error node.
+  result.n = c.newError(n.n, diag)
+  result.hasError = true
+
+# speed up debug builds by not generating stack-trace information for the
+# accessors procedures
+{.push stacktrace: off.}
+
+func kind*(x: CheckedAst): TNodeKind {.inline.} =
+  x.n.kind
+
+func len*(x: CheckedAst): int {.inline.} =
+  x.n.len
+
+func info*(x: CheckedAst): TLineInfo {.inline.} =
+  x.n.info
+
+func add*(x: var CheckedAst, n: sink PNode) {.inline.} =
+  x.hasError = x.hasError or n.kind == nkError
+  x.n.add n
+
+func add*(x: var CheckedAst, n: sink CheckedAst) {.inline.} =
+  x.hasError = x.hasError or n.hasError
+  x.n.add n.n
+
+{.pop.}
+
+template idx(len: int, x: Natural): int =
+  int(x)
+
+template idx(len: int, x: BackwardsIndex): int =
+  len - int(x)
+
+func `[]`*(x: CheckedAst, i: int|BackwardsIndex): PNode =
+  x.n[idx(x.n.len, i)]
+
+func `[]=`*(x: var CheckedAst, i: int|BackwardsIndex, n: sink PNode) =
+  x.hasError = x.hasError or n.kind == nkError
+  x.n[idx(x.n.len, i)] = n
+
+func `[]=`*(x: var CheckedAst, i: int|BackwardsIndex, n: sink CheckedAst) =
+  x.hasError = x.hasError or n.hasError
+  x.n[idx(x.n.len, i)] = n.n
+
+proc wrap*(c: ConfigRef, n: sink CheckedAst): PNode =
```

Comment:
```suggestion
proc extract*(c: ConfigRef, n: sink CheckedAst): PNode =
```

Suggesting because `CheckedAst` is itself a wrapper, so the current name is a bit confusing. I was thinking of an antonym for `prepare` that might work, but that's about the best I could come up with.